### PR TITLE
Add cmake compatibility to c-api

### PR DESCRIPTION
**Pull request by TheGreatRambler**
This PR adds CMake compatibility to the c-api, similar to that in [wasm-tools](https://github.com/bytecodealliance/wasm-tools/blob/main/crates/c-api/CMakeLists.txt). The examples are also modified to both demonstrate usage of CMake and are built by CI to verify the c-api is compiling properly.

### CODE REVIEW

**Comment by TheGreatRambler** [comment]
`error: failed to open: /home/runner/work/wasmtime/wasmtime/target/release/.cargo-lock` This PR couldn't have caused that CI error

**Comment by TheGreatRambler** [comment]
`error occurred: Failed to find tool. Is `aarch64-linux-gnu-gcc` installed?` how should I go about fixing this?

File: crates/c-api/include/wasmtime.h [hunk 1]
```diff
@@ -41,7 +57,8 @@
  *
  * * Linux - `-lpthread -ldl -lm`
  * * macOS - no extra flags needed
- * * Windows - `ws2_32.lib advapi32.lib userenv.lib ntdll.lib shell32.lib ole32.lib bcrypt.lib`
+ * * Windows - `ws2_32.lib advapi32.lib userenv.lib ntdll.lib shell32.lib
+ * ole32.lib bcrypt.lib`
```
Comment (bjorn3) [inline crates/c-api/include/wasmtime.h:61]: This wrapping makes it harder to copy the full list.

File: crates/c-api/include/wasmtime.h [hunk 2]
```diff
@@ -187,26 +204,22 @@ extern "C" {
...
  *
- * \param wat this it the input pointer with the WebAssembly Text Format inside of
- *   it. This will be parsed and converted to the binary format.
- * \param wat_len this it the length of `wat`, in bytes.
- * \param ret if the conversion is successful, this byte vector is filled in with
- *   the WebAssembly binary format.
+ * \param wat this it the input pointer with the WebAssembly Text Format inside
+ * of it. This will be parsed and converted to the binary format. \param wat_len
+ * this it the length of `wat`, in bytes. \param ret if the conversion is
+ * successful, this byte vector is filled in with the WebAssembly binary format.
```
Comment (bjorn3) [inline crates/c-api/include/wasmtime.h:210]: Can you please revert this change? I am not sure doc tools will like it. In any case it makes it harder to read.

File: crates/c-api/include/wasmtime.h [hunk 3]
```diff
@@ -187,26 +204,22 @@ extern "C" {
...
  * This function does not take ownership of `wat`, and the caller is expected to
  * deallocate the returned #wasmtime_error_t and #wasm_byte_vec_t.
  */
-WASM_API_EXTERN wasmtime_error_t* wasmtime_wat2wasm(
-    const char *wat,
-    size_t wat_len,
-    wasm_byte_vec_t *ret
-);
+WASM_API_EXTERN wasmtime_error_t *
+wasmtime_wat2wasm(const char *wat, size_t wat_len, wasm_byte_vec_t *ret);
```
Comment (bjorn3) [inline crates/c-api/include/wasmtime.h:219]: Can you please also revert this? It doesn't match the code style of the other header files.

File: examples/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+create_target(gcd gcd.c)
+create_target(hello hello.c)
+create_target(interrupt interrupt.c)
+create_target(linking linking.c)
+create_target(memory memory.c)
+create_target(multi multi.c)
+create_target(multimemory multimemory.c)
+create_target(serialize serialize.c)
+create_target(threads threads.c)
+create_target(wasi wasi/main.c)
```
Comment (bjorn3) [inline examples/CMakeLists.txt:38]: nit: missing trailing newline

File: examples/threads.c [hunk 5]
```diff
@@ -1,3 +1,27 @@
...
+function in a separate thread.
+
+You can compile and run this example on Linux with:
+
+   cargo build --release -p wasmtime-c-api
+   cc examples/threads.c \
+       -I crates/c-api/include \
+       -I crates/c-api/wasm-c-api/include \
+       target/release/libwasmtime.a \
+       -lpthread -ldl -lm \
```
Comment (bjorn3) [inline examples/threads.c:12]: nit: I believe you have to use `-pthread` instead of `-lpthread` to be portable across UNIXes.

File: crates/c-api/include/wasmtime.h (hunk 2 above)
Comment (TheGreatRambler) [reply crates/c-api/include/wasmtime.h:210]: Yeah issue with my formatter

File: examples/threads.c (hunk 5 above)
Comment (TheGreatRambler) [reply examples/threads.c:12]: That was a copy and paste from serialize.c, which uses `-lpthread`, copied so cmake could be mentioned, I could change the rest to be consistent if you'd like

**Review by bjorn3 — APPROVED** [review]
Not familiar with cmake, so someone else should take a look too.

**Comment by alexcrichton** [comment]
Thanks for this! This seems reasonable enough to me. Could this be used to replace and/or modify the existing `run-examples` crate run on CI? Additionally we ideally wouldn't add any more time to CI since building, especially in release mode, is a bit hefty. Could this ensure that the crates are only built in CI once?

**Comment by TheGreatRambler** [comment]
I've moved compilation to the `run-examples` crate and verified it worked, the CI seems stuck on my end so not sure how well it will work here

**Comment by TheGreatRambler** [comment]
CI failed due to a lack of disk space. If it becomes a problem in the future I found [this script from Apache](https://github.com/apache/flink/blob/master/tools/azure-pipelines/free_disk_space.sh) that could help, it just removes unnecessary packages from the runner.

**Comment by TheGreatRambler** [comment]
Can anyone provide any tips for why the externref example works perfectly fine on my machine but fails in the CI? I clean the build folder to save disk space, but that should only impact the `examples/build` folder, `externref.wat` isn't stored there

**Comment by TheGreatRambler** [comment]
Still running out of storage? I... I don't even know, I'm going to revert the tests and test from there

**Comment by TheGreatRambler** [comment]
Could someone (if this is possible) manually restart the CI and/or give any tips on how to fix this final issue? It has to do with `CI / Test (windows-2019)` in particular taking too long or hanging.

**Comment by alexcrichton** [comment]
I don't know CMake on Windows really at all, but that looks like some form of a legitimate loop of some kind.

File: crates/misc/run-examples/src/main.rs [hunk 6]
```diff
@@ -16,13 +16,14 @@ fn main() -> anyhow::Result<()> {
...
-        .current_dir("crates/c-api"))?;
+    println!("======== Prepare C/C++ CMake project ===========");
+    run(Command::new("cmake")
+        .arg("-Sexamples")
+        .arg("-Bexamples/build")
+        .arg("-DBUILD_SHARED_LIBS=OFF"))?;
 
     for (example, is_dir) in examples {
-        if example == "README" {
+        if example == "README" || example == "CMakeLists" || example == "build" {
```
[the remark this answers is not in the document]
Comment (TheGreatRambler) [reply crates/misc/run-examples/src/main.rs]: I pushed a solution involving `cargo read-manifest` and `cmake --build examples/build --target help`, which can be parsed to get all targets. I still need to add `--features` handling but let me know how it is

[the remark this answers is not in the document]
Comment (jameysharp) [reply crates/misc/run-examples/src/main.rs]: I like it! I had tried to find an option like `cmake --target help` but didn't search for the right things, I guess.

I see you removed the `example_to_run` command-line argument. I don't know whether anyone was using that but it seems pretty easy to keep supporting it so if you don't mind putting it back that'd be cool. It's weird that it runs two examples (the C and the Rust version) most of the time, but that seems harmless.

I think the big question now is, what will CI do with this PR this time?

[the remark this answers is not in the document]
Comment (TheGreatRambler) [reply crates/misc/run-examples/src/main.rs]: Oh whoops, added it back it. It will still check if that is a valid example before it runs, however, like the old code. I'm also interested in seeing if the CI works *for once*.

[the remark this answers is not in the document]
Comment (jameysharp) [reply crates/misc/run-examples/src/main.rs]: Yup, that looks right to me. And I'm interested too! :crossed_fingers:

File: .github/actions/free-disk-space/LICENSE [hunk 7]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
Comment (alexcrichton) [inline .github/actions/free-disk-space/LICENSE:1]: Is this github action still needed? If so I don't think we should bring in some external action for this but instead work on sharing artifacts instead of producing new ones.

File: crates/misc/run-examples/src/main.rs [hunk 8]
```diff
@@ -1,36 +1,74 @@
 use anyhow::Context;
```
Comment (alexcrichton) [inline crates/misc/run-examples/src/main.rs:1]: Is it possible to remove this crate entirely? It seems unfortunate to have both this runner script for CMake plus all the CMake support. I was hoping that this `run-examples` could be entirely deleted in favor of some CMake scripting. I don't know much about CMake myself though.

**Review by alexcrichton — COMMENTED** [review]
To confirm this isn't building more in CI now than we currently are? This PR's [linux run](https://github.com/bytecodealliance/wasmtime/runs/7440284252?check_suite_focus=true) is 10 minutes longer than `main` [last linux run](https://github.com/bytecodealliance/wasmtime/runs/7451158680?check_suite_focus=true) which could be variance but that seems pretty large.

**Comment by jameysharp** [comment]
A little over 2 minutes of that difference was the `free-disk-space` action, which makes sense to me. But the biggest difference was an extra 4+ minutes in `run-tests.sh`, which I really think shouldn't have been affected by this PR at all. Other steps were also a little longer, so I think that particular CI job runner was just more heavily loaded.

That said, I would also like to know if removing the `free-disk-space` action passes CI now. I didn't get a look at the failure mode before you added it so I'm not sure what to expect. But given that the Windows tests seem to have magically fixed themselves (right? you didn't change anything that should explain that?) I'm wondering if this PR just got extraordinarily unlucky with its CI runs.

File: crates/misc/run-examples/src/main.rs (hunk 8 above)
Comment (jameysharp) [reply crates/misc/run-examples/src/main.rs:1]: Setting aside the C examples for a second, what would that look like for the Rust examples? Some of the examples have nested Rust subprojects that need to be built first to get a .wasm blob, and I couldn't find a way to make `cargo run --example` trigger that build first. I also don't see a way to ask cargo to run all examples, just a specific one by name.

My guess is that there's still value in this run-examples crate, so I'd like to merge this PR. If we figure out how to replace what it does with a tiny shell script or something, that'd be great as another PR.

**Comment by TheGreatRambler** [comment]
`free-disk-space` was an unneccesary action I pulled in when I kept running out of space during tests. Turns out it was because I was pulling in the full rust toolchain rather than the minimal one, so it should be unneccesary.

**Comment by TheGreatRambler** [comment]
As for the Windows builds passing this time, it could be dumb luck but I actually reduced the warnings msvc reports to `/W3` (the default) just in case, it may have helped by reducing stdout.

File: .github/actions/free-disk-space/LICENSE (hunk 7 above)
Comment (TheGreatRambler) [reply .github/actions/free-disk-space/LICENSE:1]: Yeah I removed it, it was an attempt to reduce disk space usage, but I introduced a bug that had caused that disk usage and fixing it solved the problem

File: crates/misc/run-examples/src/main.rs (hunk 8 above)
Comment (TheGreatRambler) [reply crates/misc/run-examples/src/main.rs:1]: Cmake would be able to build, it would look similar to the below
```
include(ExternalProject)
ExternalProject_Add(
	wasmtime-crate
	DOWNLOAD_COMMAND ""
	CONFIGURE_COMMAND ""
	INSTALL_COMMAND "${WASMTIME_INSTALL_COMMAND}"
	BUILD_COMMAND cargo build ${WASMTIME_BUILD_TYPE_FLAG}
	BINARY_DIR ${CMAKE_CURRENT_SOURCE_DIR}
	BUILD_ALWAYS ON)
add_library(wasmtime INTERFACE)
add_dependencies(wasmtime wasmtime-crate)
```
And actually it appears you [can test](https://cmake.org/cmake/help/book/mastering-cmake/chapter/Testing%20With%20CMake%20and%20CTest.html) with cmake, I've never done it but it can be done

File: examples/README.md [hunk 9]
```diff
@@ -8,7 +8,6 @@ Each example is available in both C and in Rust. Examples are accompanied with a
 `*.wat` file which is the wasm input, or a Rust project in a `wasm` folder which
 is the source code for the original wasm file.
 
-Rust examples can be executed with `cargo run --example $name`, and C examples
-need to be compiled using your system compiler and appropriate header files.
+Rust examples can be executed with `cargo run --example $name`. C examples can be built with `mkdir build && cd build && cmake ..`. You can run `cmake --build .` to build all examples or `cmake --build . --target wasmtime-$name`, replacing the name as you wish. They can also be [built manually](https://docs.wasmtime.dev/c-api/).
```
Comment (jameysharp) [inline examples/README.md:11]: A minor style nit: Markdown in this project seems to be pretty consistently word-wrapped around 75-80 columns. Could you word-wrap this paragraph?

**Review by jameysharp — CHANGES_REQUESTED** [review]
Okay, in my opinion if you remove the `.github` changes and still CI passes, then this should be good to go. In reviewing once more I also noticed one style nit that'd be nice to clean up while you're at it.

File: crates/misc/run-examples/src/main.rs (hunk 8 above)
Comment (jameysharp) [reply crates/misc/run-examples/src/main.rs:1]: I had glanced at CTest while I was reviewing this PR earlier and at a very brief glance I didn't think it was meant for this kind of use: it seemed like it was more about running a test suite, with the usual facilities for reporting number of passing/failing tests and so on. I assumed that would be a poor fit, but I could be wrong.

I think it'd be unfortunate to require people to use CMake if they just want to try the Rust examples, so I'm skeptical about building the wasm blobs that way.

Comment (TheGreatRambler) [reply crates/misc/run-examples/src/main.rs]: I've just added integration with ctest

File: examples/README.md (hunk 9 above)
Comment (TheGreatRambler) [reply examples/README.md:11]: Oh wow that's really noticeable my bad lol

File: crates/misc/run-examples/src/main.rs (hunk 8 above)
Comment (jameysharp) [reply crates/misc/run-examples/src/main.rs]: Alright, that was simpler than I expected. :laughing: Nice!

I see this version fails on Windows because `-j4` isn't supported there. It's cool that you can run the examples in parallel but I'd also be happy without that optimization.

**Review by jameysharp — APPROVED** [review]

**Comment by jameysharp** [comment]
The new problem seems to be that the examples which rely on wasm blobs can't find them when run under CTest. The macos build is the one that failed but I think it just got to that point before any of the other job runners did, since I don't see any reason why it should be Mac-specific. Do you need to change how you set the current working directory for each test, perhaps?

**Comment by TheGreatRambler** [comment]
It has to do with the order in which I build the tests, since wasm modules are built for some c tests to run. Now the new problem is msbuild not taking the ctest arguments, and it sort of needs to accept them, without `--output-on-failure` you can't diagnose issues.

**Comment by TheGreatRambler** [comment]
Of COURSE it got stuck again

**Comment by jameysharp** [comment]
I've cancelled and restarted the windows-2019 job; 4+ hours in, it clearly wasn't going to work. I wish I had any good guesses about what's going on; I'd assumed there might be something weird in the run-examples crate, but that's gone now and we're still seeing this hang in CI. You have my sympathy for how painful this has been!

I don't know much about cmake but I see there's a [TIMEOUT property](https://cmake.org/cmake/help/latest/prop_test/TIMEOUT.html) that you can set on a test. Maybe you can set a timeout on all the tests and at least get it to fail faster? I think 30 seconds ought to be more than enough time when things are working okay.

Ooh, I see you have a new patch on the branch while I was typing this. I'll be curious to see if that fixes things.

**Comment by jameysharp** [comment]
Hey, windows-2019 got past the point where it had been stuck! Although I see my suggestion of a 30-second timeout would have been too short for `epochs-rust` and `tokio-rust`, which took closer to 120 seconds on Windows and 70 seconds on Ubuntu. So I guess 300 seconds might be a better choice, if you should decide to set timeouts.

**Comment by TheGreatRambler** [comment]
It was getting stuck because both `memory.c` and `multimemory.c` read 8 bytes into a variable on the stack rather than 4, which I guess the old testing didn't catch. Cmake catches exceptions and return codes, so it just got silently stuck. Ideally everything is fine now

**Comment by jameysharp** [comment]
Alex and I are dismayed that our CI didn't catch that bug sooner; apparently we should start using `-Werror` in CI or something. That's not a problem for this PR though, which I'm merging now. Thank you for all your effort getting this through!
